code/modules/crafting/craft.dm

```diff
@@ -0,0 +1,304 @@
+/datum/personal_crafting
+	var/busy
+	var/viewing_category = 1 //typical powergamer starting on the Weapons tab
+	var/list/categories = list(CAT_WEAPON,CAT_AMMO,CAT_ROBOT,CAT_FOOD,CAT_MISC,CAT_PRIMAL)
+	var/datum/action/innate/crafting/button
+
+/datum/personal_crafting/New(mob/living/carbon/human/H)
+	button = new()
+	button.Grant(H)
+
+/datum/personal_crafting/proc/check_contents(datum/crafting_recipe/R, list/contents)
+	main_loop:
+		for(var/A in R.reqs)
+			var/needed_amount = R.reqs[A]
+			for(var/B in contents)
+				if(ispath(B, A))
+					if(contents[B] >= R.reqs[A])
+						continue main_loop
+					else
+						needed_amount -= contents[B]
+						if(needed_amount <= 0)
+							continue main_loop
+						else
+							continue
+			return 0
+	for(var/A in R.chem_catalysts)
+		if(contents[A] < R.chem_catalysts[A])
+			return 0
+	return 1
+
+/datum/personal_crafting/proc/get_environment(mob/user)
+	. = list()
+	. += range(2, user.loc)
+	. += user.contents
+
+/datum/personal_crafting/proc/get_surroundings(mob/user)
+	. = list()
+	for(var/obj/item/I in get_environment(user))
+		if(istype(I, /obj/item/stack))
```

Comment:
If ones to add proximity checks it should be in the get_environment proc to cut out whole turfs instead of per-item checks of the same stuff